src/libsyntax/tokenstream.rs

```diff
@@ -331,35 +329,65 @@ impl TokenTree {
 /// struct itself shouldn't be directly manipulated; the internal structure is not stable,
 /// and may be changed at any time in the future. The operators will not, however (except
 /// for signatures, later on).
-#[derive(Eq,Clone,Hash,RustcEncodable,RustcDecodable)]
-pub struct TokenStream {
-    pub span: Span,
-    pub tts: Vec<TokenTree>,
+
+// #[derive(Eq,Clone,Hash,RustcEncodable,RustcDecodable)]
+#[derive(Eq,Hash,RustcEncodable,RustcDecodable)]
+pub enum TokenStream {
+  Empty,
+  Leaf    { tts : Rc<Vec<TokenTree>>, sp: Span },
+  SubLeaf { tts : Rc<Vec<TokenTree>>, offset: usize, len: usize, sp: Span },
+  Node    { left: Rc<TokenStream>, right: Rc<TokenStream>, len: usize, sp: Span},
+}
+
+
+impl clone::Clone for TokenStream {
```

Comment:
I just wanted to make sure it did the right thing; it's possible deriving works fine.